### DISCUSSION TITLE
Native Support for Navigation Bar

### DISCUSSION THREAD

**Discussion by tajmone** [pml-lang/pml-companion#16] [Ideas]
Just dropping a few ideas and thought that came to my mind regarding the default PML template...

# Home Button and NavBar

It would be nice if the default HTML template could optionally create a top horizontal NavBar, on demand, i.e. if some attributes are defined. This would allow end users to create multi-page websites or documents without having to deal with HTML and CSS.

E.g. if the `home_link=index.html` variable is defined (via CLI, etc.) then the template should display a NavBar with a "Home Page" button (e.g. with a house icon) linking to `index.html`. Additional attributes for the home link could be specified via `home_alt="Alt text"`, `home_icon=img/house.svg`, etc., according to need.

Additionally, if the `navbar` variable is non-empty, navigation menus should be added accordingly. In this case, `navbar` should be a structured variable capable of representing multiple menus, and possibly also sub-menus; something like:

`navbar.categories`, `navbar.categories.food`, `navbar.categories.fun`, `navbar.categories.fun.movies`, `navbar.categories.fun.books`, `navbar.about`, `navbar.about.info`, `navbar.about.contact`, etc.

## NavBar Settings

How could end users handle all these NavBar settings?

An obvious solution would be setting these attributes via the command line, e.g.

    pmlc --define home_link=index.html [...]
    pmlc -Dhome_link=index.html [...]

but this could soon become cumbersome, especially with websites that have many submenus.

Alternative approaches could be:

1. Hosting a header section in the PML source.
2. Using external configuration file(s)

### Meta Header Section

A PML source could optionally contain a header section before the `[doc` tag in which to define all meta-data and variables for a document.

This could be done via a YAML header, as seen in many syntaxes already:

```yaml
---
- home_link: index.html
- home_alt: Return to home page
...
```

Or, it could be delimited by a dedicated PML node, e.g. `[meta`...`meta]`, which expects key=value definitions in some simple format. Although it would make sense to have this meta section at the beginning of a document, end user could actually put it anywhere they wish, since the PML converter should handle this separately from the document contents; in this case, there could be even multiple `[meta` section per doc (i.e. split into multiple parts).

> **CONS** — The obvious downside of this approach is that it would make a document less re-usable, effectively binding it to a specific project. Still, the idea of a `[meta` section could be a handy solution for other type of metadata.

### External Configuration File(s)

Possibly, being able to define the NavBars variables in an external file would be easier for end users, and it would also allow separation of concerns and allow the same document to be used in different contexts.

There are various possible solutions at hand here, each with its pros and cons.

1. Automatically look for a same-named associated file (e.g. `home.pml` &rrarr; `home.yml`/`.nav`, etc.)
2. Specify a settings file via CLI options — e.g.

    ```
    pmlc --navbar=menus.yml
    ```

The main question here is how to manage in an easy way multiple navigation menus in complex website. If the NavBar functionality is simple, and enforces the same menus on all pages, than only a single definition file is needed; but if complex website were to support changing menus, then things might get more complicate.

Instead of having a single NavBar file for each website doc/page, it could be possible to host all definitions into a single file, by using the source filename as a reference, e.g.:

```yaml
- index.pml:
    - home_link: index.html
    - home_alt: Return to home page
    - navbar:
        - fun: fun.html
- fun.pml:
    - home_link: index.html
    - home_alt: Return to home page
    - navbar:
        - fun: fun.html
```

but since this would involve lots of redundancy, it might be better to be able to define shared settings first, and only apply overriding on a per-page basis, if a specific page contains different settings.

Probably, offering a single menus structure, shared by all pages, would be a better start-off solution, but it's always worth keeping in mind how a feature might evolve.

Also, if menu links were to point to `.pml` sources (instead of their generated `.html` files) then `pmlc` could automatically do some tracking operations on these links (e.g. don't emit the link to the current page), whereas it could assume that any `.html` links are pointing to either external websites or HTML docs generated by other tools, etc.. Asciidoctor uses this approach to both support multi-format output (i.e. it will convert `.asciidoc` to the correct extension of the current converted) as well as to distinguish when a link or reference is pointing to a project file or to an external file.

# Conclusions

Of course, the ideas exposed above are all very sketchy, the result of some quick brainstorming on some long-felt needs that I wish other markup tools offered natively (as opposed to having to hack your way through custom templates and variables).

Definitely, offering native support for a NavBar (or just the Home button) would immediately make PML usable for multi-page documents and website, even with users who don't know HTML and CSS, and wouldn't be able to create a custom template.

As any natively supported feature, all its aspects would have to be carefully considered, in order to provide a balanced solution that would scale well with complex websites, yet remain simple to use with simpler projects.

Variables definitions would definitely benefit from a YAML like structure — or using [TOML], a trimmed-down YAML-inspired alternative.

[TOML]: https://toml.io/en/ "Visit TOML website"

**Comment by pml-lang** [comment]
Good food for thought!

> Hosting a header section in the PML source.

Yes, that should be done.
If you look at the current HTML code generated by the converter, you can see that there is already a `<header>` and a `<footer>` section. Both are currently empty, but the idea is to add config parameters `html_header` and `html_footer`, which define file or URL paths to HTML code templates. These will then be integrated into the final HTML file generated by the converter. These template files can use predefined parameters provided by the converter (e.g. document title, date, etc.), as well as parameters provided by the user.

> As any natively supported feature, all its aspects would have to be carefully considered, in order to provide a balanced solution that would scale well with complex websites, yet remain simple to use with simpler projects.

Yes, native support for navbars would be even better, but needs to be "carefully considered", indeed.

> it could be delimited by a dedicated PML node

> end user could actually put it anywhere they wish

> there could be even multiple [meta section per doc (i.e. split into multiple parts).

> External Configuration File(s)

Yes, yes, yes, and yes.

Configuration must be very very flexible, in order to cover a maximum of use cases, avoid duplication, and automate as far as possible. As discussed already [here](https://github.com/pml-lang/user-manual/discussions/2#discussioncomment-517481) (slightly modified):

Configuration data can be defined as a cascade at different levels:

1. in default configuration files located in a sub-directory of PML's installation directory
2. in external, sharable resources (files or urls) located anywhere and referenced in the PML document (overrides 1)
3. in one or more configuration sections in the PML document (using PML syntax) (overrides 2)
4. as command line arguments or config files/urls referenced by a command line argument (overrides 3)

**Reply by tajmone** [reply]
> Configuration data can be defined as a cascade at different levels:

Wouldn't it be worth to also add (inserted between proposed points __1.__ and __2.__):

- the OS's user directory (`AppData/` on Win, `~` on Linux, etc.)
- the current working folder?

Although the latter might introduce some resolution problems (what's the invoking folder? The path of the builder script or that of the main PML entry file?) it might be what most shared projects are looking for — e.g. a GitHub repository could host its PML config file(s).

The former could also introduce some cross-platform resolution problems (at least, that's what I've noticed on some third party tools adopting a similar system), but in general it's possible to use it anyhow.

Also, the list could include a dedicate environment variable (e.g. `PML_CONFIG`) which (if present, non empty and containing/pointing-to a valid configuration) could interact with the command line option somehow, to allow end users to customize even when working with a project that comes with it's own toolchain.

Cascading configurations are definitely the way go here, and having all these options is key to ensure that customization is always possible (and easy) regardless of a project's settings. They are also harmless for the newby, since they don't stand in the way of normal "out-of-the-box" usage.

**Reply by pml-lang** [reply]
> also add the OS's user directory

> the current working folder

> a dedicate environment variable (e.g. PML_CONFIG)

Good point! Yes, these must also be added in the cascade.
	
> cross-platform resolution problems

This shouldn't be a problem because Java manages this in a platform-compatible way.

**Reply by tajmone** [reply]
> This shouldn't be a problem because Java manages this in a platform-compatible way.

I remember that in [Highlight]  (a syntax highlighter project project which I've been collaborating for many years) the difference between Windows and Linux/Mac, in this respect, was that Linux and Mac add an additional look up directory compared to Windows.

I can't quite remember the details and reasons for this, but this is the table from Highlight's Wiki _[Highlight Configuration-Search Paths]_:


|        Linux/Mac        |        Windows        |                                     |
|-------------------------|-----------------------|-------------------------------------|
| `~/.highlight/`         |                       |                                     |
| `--data-dir`            | `--data-dir`          | path of `--data-dir` option         |
| `$HIGHLIGHT_DATADIR`    | `%HIGHLIGHT_DATADIR%` | path of `HIGHLIGHT_DATADIR` env var |
| `/usr/share/highlight/` | Highlight directory   | path of Highlight executable        |
| `/etc/highlight/`       |                       | location of "`filetypes.conf`"      |
| current directory       | current directory     | (fallback)                          |

For some reason, on Windows, Linux's `~/.highlight/` wasn't being looked up — which should probably map to `%HOMEPATH%\.highlight`  (usually `C:\Users\<username>\.highlight`), although I'm not sure if that's the corresponding Win path for it.

But my memory of the problem is a bit faded, it's been a few years since I looked into it, and I can't remember if it had something to with C++'s Boost library, or whatever other reason, I just remember that in one particular occasion it did create some problems with a specific project where I was trying to override some predefined settings with my own.

I think there was also a problem with the order in which these were being looked up in the different OSs.

I'm just saying that it's worth keeping an eye on these differences, especially when relying on some library to handle this.

### Reference Links

- [§3.8. CONFIG FILE SEARCH]
- [Highlight Wiki » Custom Data Folders]



[§3.8. CONFIG FILE SEARCH]: https://gitlab.com/saalen/highlight/-/blob/master/README.adoc#user-content-config-file-search
[Highlight]: https://gitlab.com/saalen/highlight/

[Highlight Wiki » Custom Data Folders]: https://gitlab.com/saalen/highlight/-/wikis/Custom-Data-Folders#highlight-configuration-search-paths
[Highlight Configuration-Search Paths]: https://gitlab.com/saalen/highlight/-/wikis/Custom-Data-Folders#highlight-configuration-search-paths

**Reply by pml-lang** [reply]
> worth keeping an eye on these differences

Indeed. Even in Java's "Write once, run anywhere" world, surprises are never excluded.

Thanks for the info and links.

**Reply by celtic-coder** [reply]
> For some reason, on Windows, Linux's `~/.highlight/` wasn't being looked up — which should probably map to `%HOMEPATH%\.highlight` (usually `C:\Users\<username>\.highlight`), although I'm not sure if that's the corresponding Win path for it.

Technically what you require is `%homedrive%%homepath%`

At a CMD prompt, type: `echo %homedrive%%homepath%` to see your own settings. To see all the related environment variables, use `set home`
